contracts/src/registry/ETHRegistrar.sol

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+import {IETHRegistrar} from "./IETHRegistrar.sol";
+import {IRegistry} from "./IRegistry.sol";
+import {IERC1155Singleton} from "./IERC1155Singleton.sol";
+import {IETHRegistry} from "./IETHRegistry.sol";
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+import {IPriceOracle} from "./IPriceOracle.sol";
+import {ERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import {NameUtils} from "../utils/NameUtils.sol";
+
+contract ETHRegistrar is IETHRegistrar, AccessControl {
+    bytes32 public constant CONTROLLER_ROLE = keccak256("CONTROLLER_ROLE");
+    uint256 public constant MIN_REGISTRATION_DURATION = 28 days;
+    uint64 private constant MAX_EXPIRY = type(uint64).max;
+
+    error MaxCommitmentAgeTooLow();
+    error UnexpiredCommitmentExists(bytes32 commitment);
+    error DurationTooShort(uint64 duration);
+    error CommitmentTooNew(bytes32 commitment);
+    error CommitmentTooOld(bytes32 commitment);
+    error NameNotAvailable(string name);
+    error InsufficientValue();
+
+    IETHRegistry public immutable registry;
+    IPriceOracle public immutable prices;
+    uint256 public immutable minCommitmentAge;
+    uint256 public immutable maxCommitmentAge;
+
+    mapping(bytes32 => uint256) public commitments;    
+
+    constructor(address _registry, IPriceOracle _prices, uint256 _minCommitmentAge, uint256 _maxCommitmentAge) {
+        registry = IETHRegistry(_registry);
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        if (_maxCommitmentAge <= _minCommitmentAge) {
+            revert MaxCommitmentAgeTooLow();
+        }
+
+        prices = _prices;
+        minCommitmentAge = _minCommitmentAge;
+        maxCommitmentAge = _maxCommitmentAge;
+    }
+
+    /**
+     * @dev Check if a name is valid.
+     * @param name The name to check.
+     * @return True if the name is valid, false otherwise.
+     */
+    function valid(string memory name) public pure returns (bool) {
+        return bytes(name).length >= 3;
+    }
+
+    /**
+     * @dev Check if a name is available.
+     * @param name The name to check.
+     * @return True if the name is available, false otherwise.
+     */
+    function available(string calldata name) external view returns (bool) {
+        address subregistry = address(registry.getSubregistry(name));
+        return subregistry == address(0);
```

Comment:
We want to permit users to set their subregistry to 0 for valid names if they don't have subdomains. We should probably check the expiration date, instead.